### ISSUE TITLE
Idea on how to continue this app

### ISSUE THREAD

**Issue by ethanaobrien** [kzahel/web-server-chrome#271]
Hey @kzahel There is one way we could continue this app. Its a little heavy on storage (200-300 mb) but I think that it is a good idea.
We could use [the chromium archives](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html) to download an older version of chrome (I would do around 75) and then we could create a bat (Or sh for linux) to use chrome.exe (or just chrome, for linux) command line flags to load the unpacked app and to start the unpacked app. I've already gotten the app to load unpacked. I am trying to use the chrome.exe to start the app. Again, This wouldnt be a bad idea (And is extremely easy) just a little heavy on storage.

**Comment by terreng** [comment]
@ethanaobrien @kzahel 

Hi there. I've been using this Chrome app for years, and so I was sad to learn that it's going away soon. I really liked the simplicity of it, and I used it a lot.

I decided to build a new app inspired by this one using Electron and [send](https://www.npmjs.com/package/send). It's a standalone app for Windows or macOS (this is similar to what you described as option 1).

It works very similarly to Web Server for Chrome with some of the same features. But in addition, my version lets you run more than one web server at the same time.

<img width="531" alt="Screen Shot 2021-07-13 at 8 21 25 PM" src="https://user-images.githubusercontent.com/11605395/125555825-ad02d8c3-8d09-49ea-8547-25f0916c397e.png">

It's open source (MIT license). Take a look: https://github.com/terreng/web-server
To install: https://github.com/terreng/web-server/releases

It's been working great for me, so I thought I'd share it here.

It's very simple at the moment, and it doesn't have some of the features this extension did.
If you are interested, I would appreciate your help developing it further.

**Comment by ethanaobrien** [comment]
@terreng Thats really cool! I've tried converting to an electron app (And never got past the server part) and I took a glimpse at your code (I'll look deeper later) but [here](https://github.com/terreng/web-server/blob/main/index.js#L152) it looks like we could just transform the request into something that the handler can understand and then we wouldn't have to rewrite ANY of the code.
I have 1 question that I ask you to answer
How does the filesystem work. I think this is the hardest part - translating the `fs.getByPath()` function

This is really cool. I'm going to take a look when I have more time and maybe work on converting it (hopefully I don't have to re-write everything I did in #264 - But I'll do it to keep this app alive)!

**Comment by terreng** [comment]
> How does the filesystem work. I think this is the hardest part - translating the `fs.getByPath()` function

It runs in the Node environment, so it uses the [node file system api](https://nodejs.org/api/fs.html). It uses [send](https://www.npmjs.com/package/send) to serve files from the chosen directory, but that could be replaced with a custom solution to add more features. Let me know if that answers your question. 

It would be great if you were able to drop in the server code from this repo into the Electron program. I built the UI without any frameworks, because they're not necessary for such a simple program. I can work on the UI side of things, and it sounds like you might be able to take the lead on the server part.

**Comment by ethanaobrien** [comment]
(I send this message assuming it makes it through) Ok, cool. I can do the server part and you can do the UI. I'm not very good with making a ui, I'm better at the actual code. I can't do anything today but tomorrow I can probably start to convert it. I plan to have all the options currently supported (there are some I'm still working on) plus an option to have http2. And yes, the file system part answers my question. Do you want to create a repository called something like web-server-chrome-conversion or just a repository where we can work on converting the app? (If you do so, please don't forget to share it with me) Let me know what you decide!

**Comment by terreng** [comment]
> Do you want to create a repository called something like web-server-chrome-conversion or just a repository where we can work on converting the app? (If you do so, please don't forget to share it with me) Let me know what you decide!

Do you want to just use [the repository I created](https://github.com/terreng/web-server/tree/main) as a starting point and then start adding this server code to it?

**Comment by ethanaobrien** [comment]
Sure, that sounds good. I do have a questions.
How does the server read the options? This part would be difficult because the handler reads the options from `app.opts.optSomeOpt` and `document.getElementById()`can become a pain

Also, can you start by building the UI to list every option. Please use my pull request to see the options. Under pull requests, it's called 'feature updates and bug fixes'

**Comment by terreng** [comment]
> How does the server read the options? 

server.js reads the config.json file into a variable when the program starts, and passes it on to main.js (which is running inside the browser instance) using node ipc. Then, any time the user changes settings, main.js sends a message back to server.js to update the config file and update its variable. So the server can read all the settings through the variable config instantly without needing to check the current state of any of the checkboxes. Let me know if that makes sense.

> Also, can you start by building the UI to list every option.

Can you get started with the options that I have added so far? You can also manually update the config.json stored on your computer to add new settings. They will get loaded in and read by server.js even though they don't show up on the UI yet.

I see your pull request, and it would actually be helpful if you could open an issue on my repository with a list of the options that need to be added to the UI.

**Comment by ethanaobrien** [comment]
Yeah, that all makes sense. The part about being able to add more options in config.json is really helpful.
Here is my plan:
1. Convert the base handler
2. Convert the filesystem 
3. Plug in the entry handler and fix bugs as needed
4. Give you a list of options that will need to be implemented in the ui
5. Do a Release

Does this plan sound good?
Also, how should we release? - I think we should (before we merge into kzahel's repository, if we merge) to have him do a release that adds a card on the Ui saying that chrome apps are deprecated and as a result this is now an electron app and have a link to the app.

**Comment by terreng** [comment]
I'm glad you are interested in taking the lead on converting features from this project into the new one. You should work on it in whatever order is best for you. Let me know how I can help.

It would be up to kzahel to decide if the new version should be a continuation of web server for chrome, and he'd have to give permission to use the same logo and add a link to the new version from the old one.

Alternatively, we could make the new version a separate project, with a new name / logo. 
Here's what I have right now:

<img width="144" alt="Screen Shot 2021-07-14 at 4 38 06 PM" src="https://user-images.githubusercontent.com/11605395/125706126-9446712f-2d8a-4167-a631-afe3be2a96c3.png">

We can do whatever works best.

**Comment by kzahel** [comment]
Cool project! Feel free to use any logos or assets or branding materials as you please. (though web server for chrome doesn't make much sense as a name as it's not really "for chrome" anymore as an electron app)

**Comment by ethanaobrien** [comment]
Cool! I'll get to work when I can!

**Comment by ethanaobrien** [comment]
@terreng I need some help. Seems as if send takes the file and writes it, which will not work for what we are trying to do. Do you have any other suggestions on how we could read the file. Reading the file and writing the file need to be separate

**Comment by terreng** [comment]
> Seems as if send takes the file and writes it

That's right, `send` takes a path and returns a file stream. I would suggest [taking a look at how it works](https://github.com/pillarjs/send/blob/master/index.js), and implementing it yourself. You can get rid of the send package, and move its code to the server.js file. That way you can change whatever you want.

**Comment by ethanaobrien** [comment]
@terreng Do you think using plain node `fs` module would work. I know it supports reading/writing/deleting, everything we would need.

**Comment by terreng** [comment]
> Do you think using plain node `fs` module would work. I know it supports reading/writing/deleting, everything we would need.

Yeah, `fs` can do everything we'll need. When it comes to reading files, [we'll want to use streams](https://github.com/pillarjs/send/blob/de073ed3237ade9ff71c61673a34474b30e5d45b/index.js#L796). This speeds things up for big files, because it means we can start serving the file while it is still being read, instead of waiting for the entire file to finish being read.

**Comment by ethanaobrien** [comment]
Ok, sounds cool!

**Comment by ethanaobrien** [comment]
@terreng News: I have converted the base handler (Yaaay!). Problem: I cant find an easy way to use fs. Do you know of a way to use fs to get a path and have it return directory or file or something like that. Once I get the file system converted, 90% of the handler will snap into place. I just need some htlp finding the right command

**Comment by terreng** [comment]
> Do you know of a way to use fs to get a path and have it return directory or file or something like that

I'm not entirely sure what you mean. Could you clarify what you're trying to do?

**Comment by ethanaobrien** [comment]
Yes, sorry. I should've included more detail. I'm having trouble translating the getByPath function. I cannot find a fs function that will retrieve info on, if it is a directory, the directory, and if it is a file, a file. Basically, what would be the best way to retrieve the info on the path, then from here, I could either get directory contents or get the file, basically what fs function will give me this info. If I'm still unclear, just let me know.

**Comment by terreng** [comment]
You'll want to use [fs.stat](https://nodejs.org/api/fs.html#fs_fs_stat_path_options_callback) to get information about a path.
```javascript
fs.stat(path, function(error, stats) {
    console.log(stats.isDirectory())
})
```

Use [fs.readdir](https://nodejs.org/api/fs.html#fs_fs_readdir_path_options_callback) to get files in a directory (although I'm not sure you'll need to do this). 

Use [fs. createReadStream](https://github.com/pillarjs/send/blob/de073ed3237ade9ff71c61673a34474b30e5d45b/index.js#L796) to read a file.

Let me know if this will enable you to achieve what you're trying to do.

**Comment by ethanaobrien** [comment]
(Sorry if my last message came in twice)
Thank you, this will help me put together most of the file system. I do have a few more things that I am going to go ahead and ask.
1. I plan to have a directory listing. Will the feature that you told me about that gets the directory contents return the size and the date modified? If not, how can I get the size and date modified?
2. How can I delete a file (for delete requests)
3. How can I write a file (for our requests)

Thank you

**Comment by terreng** [comment]
> 1. I plan to have a directory listing. Will the feature that you told me about that gets the directory contents return the size and the date modified? If not, how can I get the size and date modified?

Yes, it should provide that information.

> 2. How can I delete a file (for delete requests)

Remove a file using [fs.unlink](https://nodejs.org/api/fs.html#fs_fs_unlink_path_callback).

> 3. How can I write a file (for our requests)

Use [fs.writeFile](https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback) to overwrite the contents of a file.

**Comment by ethanaobrien** [comment]
@terreng I have about 50% of the app converted. Haven't started with the PUT/DELETE/POST requests because the GET request is a little buggy still. Do you want me to go ahead and give you a list of options so you can start pulling them into the UI?

**Comment by terreng** [comment]
It would be great if you could open an issue on my repo with the list of the options that I need to add.

**Comment by terreng** [comment]
@kzahel Do you still have the original version of the web server for chrome logo? If so, could you share it?

**Comment by kzahel** [comment]
You mean an SVG? I just looked and couldn't find it. Maybe it's on some
backup somewhere but most likely I lost it

**Comment by terreng** [comment]
> You mean an SVG?

Yeah, I was hoping you might have something vector. I was just trying to get an icon a little sharper. No worries if it's lost though. Thanks for checking.

**Comment by ethanaobrien** [comment]
@kzahel I also have a question. Would you mind if in my pull request if I added a react Ui card saying that chrome apps are going away and that this app has been converted?

**Comment by kzahel** [comment]
Yeah for sure.

**Comment by terreng** [comment]
> chrome apps are going away

@ethanaobrien How long until chrome apps stop working?

**Comment by ethanaobrien** [comment]
I don't know. They've announced I think earlier this month that you can no longer upload private chrome apps. The apps button also does not show in the chrome web store anymore. Only extensions is there.

**Comment by its-pablo** [comment]
@terreng sry for being late to the conversation, but where does it say this app is going away? I've looked around and couldn't find it

**Comment by terreng** [comment]
@its-pablo Chrome apps are being discontinued, so all Chrome apps including this one are going away. A replacement program for this app, [Simple Web Server](https://github.com/terreng/simple-web-server/), will be released very soon, at which time a message will be added to the Web Server for Chrome extension directing people to the new app. I hope this answers your question.
